game-core/src/main/java/games/strategy/triplea/delegate/ScrambleLogic.java

```diff
@@ -0,0 +1,223 @@
+package games.strategy.triplea.delegate;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitCollection;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.TripleAUnit;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.IBattle.BattleType;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.PredicateBuilder;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.util.Tuple;
+
+/**
+ * ScrambeLogic encapsulates the logic for finding possible units that can scramble to defend a
+ * given territory.
+ */
+public class ScrambleLogic {
+  private final GameData data;
+  private final PlayerId player;
+  private final Set<Territory> territoriesWithBattles;
+  private BattleTracker battleTracker;
+  @Getter private final Predicate<Unit> airbaseThatCanScramblePredicate;
+  private final Predicate<Territory> canScrambleFromPredicate;
+  private final int maxScrambleDistance;
+
+  public ScrambleLogic(
```

Comment:
Would you mind pointing out in this class any new/updated code that was not purely moved? It looks like this is just a straight move and a new constructor on top. I'd like to be sure the new additions/updates are reviewed a bit more closely.

Second, is this extracted code in a better position to be tested?

Reply:
The logic remains unchanged (no functional changes), but some things moved around. For example, previously the map was keyed by both "to" territory and defending player. Now, the defending player is no longer part of the key (it was 1-1 to the "to" territory) and is instead computed afterwards by caller.

Reply:
As for testing, it's certainly the case that having smaller helper classes makes it easier to test than the previous state.